### PR TITLE
feat(map): add pointcloud divider and pointcloud merger

### PR DESCRIPTION
**Pull request by anhnv3991**
## Description

- Added pointcloud divider and pointcloud merger with modified classes for reading/writing PCD with minimize memory requirements
- Reworked parameters: move parameters to YAML file and added schema.json files
- Added launch files (py and xml)

## Tests performed

- To test the pointcloud divider, run the below command.
```
$ ros2 launch autoware_pointcloud_divider pointcloud_divider.launch.xml input_pcd_or_dir:=<path_to_input_PCD_folder_or_PCD_file> output_pcd_dir:=<path_to_output_PCD_folder> prefix:=<prefix_for_output_PCDs>
```
- To test the pointcloud merger, run the below command.
```
$ ros2 launch autoware_pointcloud_merger pointcloud_merger.launch.xml input_pcd_dir:=<path_to_the_input_PCD_folder> output_pcd:=<path_to_the_merged_PCD_file>
```
- The memory consumption should not exceed 3GB
## Notes for reviewers
- Currently these tools only support XYZ or XYZI PCD files. Other format like XYZRGB are not supported.
- The default point type for readding/writing PCD files is pcl::PointXYZI. To change the point type please go to the YAML file at config/ and change point_type: to `"point_xyz"` or `"point_xyzi"` (string type).
- Other parameters (like the segmentation resolution or downsampling resolution) are also specified via YAML files.

## Interface changes

- Parameters are all moved to YAML file and tools are launched via `ros2 launch` instead of shell scripts like MapIV's tools.

## Effects on system behavior

None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by anhnv3991** [comment]
@SakodaShintaro Thank you for reviewing the point cloud divider.
Regarding this problem

> I tried pointcloud_divider to [nishishinjuku_map](https://tier4.github.io/AWSIM/GettingStarted/QuickStartDemo/#launching-autoware).
> It was very slower than the old pointcloud_divider.
> Is there any big difference from the old one?

I found a problem with the code. The file you shared is a bit different from usual PCD files. Its header look like this
```
FIELDS intensity _ x y z _
SIZE 4 1 4 4 4 1
TYPE F U F F F U
COUNT 1 12 1 1 1 4

```
A typical PCD's header often looks like this
```
FIELDS x y z intensity
SIZE 4 4 4 4
TYPE F F F F
COUNT 1 1 1 1
```
I don't know how your file was created but the order of point fields was changed, and there is some padding bytes (fields "_"). My PCD reader did not expected those thing previously. I modified the code to address those fields, and the code should work now.

File: map/autoware_pointcloud_divider/README.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# autoware_pointcloud_divider
+
+(Updated 2024/08/16)
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/README.md:3]:
```suggestion
```
not need

File: map/autoware_pointcloud_divider/README.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+This is a tool for processing pcd files, and it can perform the following functions:
+
+- Dividing point clouds
+- Downsampling point clouds
+- Generating metadata to efficiently handle the divided point clouds
+
+## Supported Data Format
+
+**Currently, only `pcl::PointXYZ` and `pcl::PointXYZ` are supported. Any PCD will be loaded as those two types.**
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/README.md:13]:
```suggestion
**Currently, only `pcl::PointXYZ` and `pcl::PointXYZI` are supported. Any PCD will be loaded as those two types.**
```

File: map/autoware_pointcloud_divider/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+- When loading `XYZ`-only data, the `intensity` field is assigned 0.
+
+## Installation
+
+```bash
+cd <PATH_TO_pilot-auto.*> # OR <PATH_TO_autoware>
+cd src/
+git clone git@github.com:autowarefoundation/autoware_tools.git
+cd ..
+colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release --catkin-skip-building-tests --symlink-install --packages-up-to pointcloud_divider
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/README.md:27]:
```suggestion
colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release --catkin-skip-building-tests --symlink-install --packages-up-to autoware_pointcloud_divider
```

File: map/autoware_pointcloud_divider/launch/pointcloud_divider.launch.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# Copyright 2022 Tier IV, Inc. All rights reserved.
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/launch/pointcloud_divider.launch.py:1]:
```suggestion
# Copyright 2024 TIER IV, Inc. All rights reserved.
```

File: map/autoware_pointcloud_merger/README.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# autoware_pointcloud_merger
+
+(Updated 2024/08/16)
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_merger/README.md:3]:
```suggestion
```
not need

File: map/autoware_pointcloud_divider/docs/how_to_be_downsampled.drawio.svg
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/docs/how_to_be_downsampled.drawio.svg:1]: update figure.
`merge_pcds` does not exist

File: map/autoware_pointcloud_divider/docs/output_file_name_pattern.drawio.svg
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/docs/output_file_name_pattern.drawio.svg:1]: update figure.
merge_pcds does not exist

File: map/autoware_pointcloud_merger/README.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+- When loading `XYZ`-only data, the `intensity` field is assigned 0.
+
+## Installation
+
+```bash
+cd <PATH_TO_pilot-auto.*> # OR <PATH_TO_autoware>
+cd src/
+git clone git@github.com:autowarefoundation/autoware_tools.git
+cd ..
+colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release --catkin-skip-building-tests --symlink-install --packages-up-to pointcloud_merger
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_merger/README.md:26]:
```suggestion
colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release --catkin-skip-building-tests --symlink-install --packages-up-to autoware_pointcloud_merger
```

File: map/autoware_pointcloud_merger/launch/pointcloud_merger.launch.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
+# Copyright 2022 Tier IV, Inc. All rights reserved.
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_merger/launch/pointcloud_merger.launch.py:1]:
```suggestion
# Copyright 2024 TIER IV, Inc. All rights reserved.
```

File: map/autoware_pointcloud_divider/src/pcd_divider.cpp [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+  }
+
+  saveTheRest();
+
+  RCLCPP_INFO(logger_, "Merge and downsampling... ");
+
+  // Now merge and downsample
+  mergeAndDownsample();
+
+  std::string yaml_file_path = output_dir_ + "/" + file_prefix_ + "_metadata.yaml";
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/src/pcd_divider.cpp:139]:
```suggestion
  std::string yaml_file_path = output_dir_ + "/" + "pointcloud_map_metadata.yaml";
```

The metadata file name is recommended to be "pointcloud_map_metadata.yaml," so I would like to ensure it is always saved with that name.

https://github.com/autowarefoundation/autoware.universe/tree/main/map/map_loader#directory-structure-of-these-files

File: map/autoware_pointcloud_divider/src/pcd_divider.cpp [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+  util::remove(dir_path);
+}
+
+template <class PointT>
+std::string PCDDivider<PointT>::makeFileName(const GridInfo<2> & grid) const
+{
+  std::string file_name = output_dir_;
+
+  file_name = file_name + file_prefix_ + "_";
+  file_name = file_name + std::to_string(grid.ix) + "_" + std::to_string(grid.iy) + ".pcd";
```
Comment (YamatoAndo) [inline map/autoware_pointcloud_divider/src/pcd_divider.cpp:431]:
```suggestion
  std::string file_name = file_prefix_ + "_" + std::to_string(grid.ix) + "_" + std::to_string(grid.iy) + ".pcd";
```
not need output_dir name.
https://github.com/autowarefoundation/autoware.universe/tree/main/map/map_loader#metadata-structure

Comment (anhnv3991) [reply map/autoware_pointcloud_divider/src/pcd_divider.cpp]: @YamatoAndo I removed that output_dir_.

**Review by YamatoAndo — APPROVED** [review]

**Pull request merged**